dcrec/secp256k1/signature.go

```diff
@@ -261,137 +263,246 @@ func (sig *Signature) IsEqual(otherSig *Signature) bool {
-		return nil, errors.New("signature S is negative")
-	case errExcessivelyPaddedValue:
-		return nil, errors.New("signature S is excessively padded")
-	}
-	signature.s = new(big.Int).SetBytes(sBytes)
-	index += sLen
-
-	// sanity check length parsing
-	if index != len(sigStr) {
-		return nil, fmt.Errorf("malformed signature: bad final length %v != %v",
-			index, len(sigStr))
-	}
-
-	// Verify also checks this, but we can be more sure that we parsed
-	// correctly if we verify here too.
-	// FWIW the ecdsa spec states that R and S must be | 1, N - 1 |
-	// but crypto/ecdsa only checks for Sign != 0. Mirror that.
-	if signature.r.Sign() != 1 {
-		return nil, errors.New("signature R isn't 1 or more")
-	}
-	if signature.s.Sign() != 1 {
-		return nil, errors.New("signature S isn't 1 or more")
-	}
-	if signature.r.Cmp(curve.Params().N) >= 0 {
-		return nil, errors.New("signature R is >= curve.N")
-	}
-	if signature.s.Cmp(curve.Params().N) >= 0 {
-		return nil, errors.New("signature S is >= curve.N")
-	}
-
-	return signature, nil
-}
-
 // ParseDERSignature parses a signature in the Distinguished Encoding Rules
-// (DER) format per section 10 of [ISO/IEC 8825-1].
-func ParseDERSignature(sigStr []byte) (*Signature, error) {
-	return parseSig(sigStr)
-}
+// (DER) format per section 10 of [ISO/IEC 8825-1] and enforces the following
+// additional restrictions specific to secp256k1:
+//
+// - The R and S values must be in the valid range for secp256k1 scalars:
+//   - Negative values are rejected
+//   - Zero is rejected
+//   - Values greater than or equal to the secp256k1 group order are rejected
+func ParseDERSignature(sig []byte) (*Signature, error) {
+	// The format of a DER encoded signature for secp256k1 is as follows:
+	//
+	// 0x30 <total length> 0x02 <length of R> <R> 0x02 <length of S> <S>
+	//   - 0x30 is the ASN.1 identifier for a sequence
+	//   - Total length is 1 byte and specifies length of all remaining data
+	//   - 0x02 is the ASN.1 identifier that specifies an integer follows
+	//   - Length of R is 1 byte and specifies how many bytes R occupies
+	//   - R is the arbitrary length big-endian encoded number which
+	//     represents the R value of the signature.  DER encoding dictates
+	//     that the value must be encoded using the minimum possible number
+	//     of bytes.  This implies the first byte can only be null if the
+	//     highest bit of the next byte is set in order to prevent it from
+	//     being interpreted as a negative number.
+	//   - 0x02 is once again the ASN.1 integer identifier
+	//   - Length of S is 1 byte and specifies how many bytes S occupies
+	//   - S is the arbitrary length big-endian encoded number which
+	//     represents the S value of the signature.  The encoding rules are
+	//     identical as those for R.
+	//
+	// NOTE: The DER specification supports lengths of more than 1 byte,
+	// however, since this is specific to secp256k1 signatures, all lengths
+	// under will be a single byte.
```

Comment:
It's referring to the length of the count, not of the R or S.  For our purposes, this count will always will fit into a single byte, instead of needing a multi-byte compact representation for longer R and S fields.

Reply:
Oh, ok. Just taking out the work "under" makes sense then.